### ISSUE TITLE
Allow renaming of uploaded artifacts

### ISSUE THREAD

**Issue by keithbranton** [buildkite-plugins/artifacts-buildkite-plugin#25]
I'm uploading artifacts that are nested a few layers deep in my maven project, and it would be much easier to deal with them in subsequent steps if I could rename them (essentially remove the directories) as I upload them. I see the buildkite agent supports this, but can't see any way to do it using the plugin. Can this please be added?

**Comment by toolmantim** [comment]
That's a great suggestion @keithbranton. Any idea on how you'd like the config syntax to be, based on your maven project?

**Comment by keithbranton** [comment]
Not sure - the yaml string for the upload could include some sort of delimiter and a destination. If nobody ever used spaces in filenames I'd vote for that, but maybe something like -> would be safer.
```
  plugins:
    ecr#v1.1.4:
      login: true
      account_ids: "880807510442"
      no-include-email: true
      region: us-east-2
    ssh://git@github.com/alianza-dev/maven-buildkite-plugin.git#v1.1.0:
      job: release
    artifacts#v1.2.0:
      upload:
      - rmq-shovel/target/rmq-shovel.jar -> rmq-shovel.jar
      - rmq-shovel/target/classes/build.properties -> build.properties
      - rmq-shovel/target/classes/config.yml -> config.yml
      - outbound-to-rabbitmq/target/outbound-to-rabbitmq.jar -> outbound-to-rabbitmq.jar
```
or perhaps it would be better to support objects - in this case it would be helpful to be able to support simple strings as well for what is possible the common case (and backward compatibility)
```
      upload:
      - src: rmq-shovel/target/rmq-shovel.jar
        target: rmq-shovel.jar
      - src: rmq-shovel/target/classes/build.properties
        target: build.properties
      - rmq-shovel/target/classes/config.yml
```
another possibility would be to include another optional section
```
      upload:
      - rmq-shovel/target/rmq-shovel.jar
      - rmq-shovel/target/classes/build.properties
      destination:
      - rmq-shovel.jar
      - build.properties
```
Looking at all these, I'd favor the second option - it seems like the most intentional - but maybe this will spark a better idea?

**Comment by toolmantim** [comment]
Thanks @keithbranton!

Just to be clear for future travellers, with the [buildkite-agent upload](https://buildkite.com/docs/agent/v3/cli-artifact#uploading-artifacts) command we have this example in our docs, if you want to strip the directory:

```shell
cd log
buildkite-agent artifact upload test.log
```

That's not possible with this plugin (yet).

I like your second suggestion too @keithbranton.

I wonder how the  would work with glob patterns also?

```yml
steps:
  - command: make
    plugins:
      artifacts#v1.2.0:
        upload:
          - src: rmq-shovel/target/classes/build.properties
            target: build.properties
          - src: **/*.jar
            target: .
  - wait
  - command: script/test-jar.sh
    plugins:
      artifacts#v1.2.0:
        download:
          - src: app.jar
            target: .
```

🤔

Related there's also buildkite/agent#752 and the super long-running buildkite/agent#91 and buildkite/agent#186, where @keithpitt has thrown together some other options.

But this plugin is a good place to experiment with a new syntax.

**Comment by keithbranton** [comment]
@toolmantim 

These are both cool ideas, if they are easy to implement, and I'd probably use the upload globs if it was available. If you are aware of others who would like the uploading/renaming with globs, and being able to relocate downloads - then great, but I will think you are just as awesome if I can just get relocating individual upload files :)

if buildkite/agent#752 was implemented and there was a way to enable that option in the upload plugin - that would be a fine alternative for my use cases.

**Comment by pzeballos** [comment]
Closing this because it was fixed on PR #27
